runtime/compiler/optimizer/J9ValuePropagation.cpp

```diff
@@ -613,6 +613,11 @@ static TR_YesNoMaybe isValue(TR::VPConstraint *constraint)
    TR::Compilation *comp = TR::comp();
    TR_OpaqueClassBlock *clazz = type->getClass();
 
+   if (TR::Compiler->cls.isClassHasIdentity(clazz))
+      {
+      return TR_no;
+      }
+
```

Comment:
There is a [test below of `isClassArray`](https://github.com/eclipse-openj9/openj9/blob/974bdd6c51cb71fab4ea4ab50b5fab019141577d/runtime/compiler/optimizer/J9ValuePropagation.cpp#L626-L631).  It's my understanding that array classes will be marked as having identity, so that special check for `isClassArray` can be removed - but please test that out first, just to be safe!

Reply:
In my experiment, I found there are cases where `isClassArray` is `true` but `classHasIdentity` is  `false`. Checking on `isClassArray` should be kept until the `romClass` modifiers  is set up accordingly

```
 !j9class 0x4b800
..
	0x24: U32 classFlags = 0x00000000 (0)
 ..
!j9romclass 0x00007F5E65261260
..
	0x10: U32 modifiers = 0x00010411 (66577)
..
#define J9AccClassHasIdentity 0x20

#define J9AccClassArray 0x10000
```

```
> !j9class 0x4b800
J9Class at 0x4b800 {
  Fields for J9Class:
	0x0: UDATA eyecatcher = 0x0000000099669966 (2573637990)
	0x8: struct J9ROMClass* romClass = !j9romclass 0x00007F5E65261260
	0x10: void** superclasses = !j9x 0x000000000004B790
	0x18: UDATA classDepthAndFlags = 0x0000000080070001 (2147942401)
	0x20: U32 classDepthWithFlags = 0x00000000 (0)
	0x24: U32 classFlags = 0x00000000 (0)
	0x28: struct J9ClassLoader* classLoader = !j9classloader 0x00007F5E600988D8
	0x30: struct J9Object* classObject = !j9object 0x00000000A2E20B60 // java/lang/Class
	0x38: volatile UDATA initializeStatus = 0x0000000000000001 (1)
	0x40: struct J9Method* ramMethods = !j9method 0x000000000004B600 // <FAULT>
	0x48: UDATA* ramStatics = !j9x 0x0000000000000001
	0x50: struct J9Class* arrayClass = !j9class 0x00000000001F2A00 // [[I
	0x58: UDATA totalInstanceSize = 0x000000000004B600 (308736)
	0x60: struct J9ITable* lastITable = !j9itable 0x0000000000049B90
	0x68: UDATA* instanceDescription = !j9x 0x0000000000000001
	0x70: UDATA* instanceLeafDescription = !j9x 0x0000000000000001
	0x78: UDATA instanceHotFieldDescription = 0x0000000000000000 (0)
	0x80: UDATA selfReferencingField1 = 0x0000000000000000 (0)
	0x88: UDATA selfReferencingField2 = 0x0000000000000000 (0)
	0x90: struct J9Method* initializerCache = !j9method 0x0000000000000000
	0x98: UDATA romableAotITable = 0x00007F5E5EC83830 (140043293833264)
	0xa0: UDATA packageID = 0x00007F5E600988D8 (140043314890968)
	0xa8: struct J9Module* module = !j9module 0x00007F5E60099A18
	0xb0: struct J9Class* subclassTraversalLink = !j9class 0x000000000004B400 // [S
	0xb8: struct J9Class* subclassTraversalReverseLink = !j9class 0x000000000004BC00 // [J
	0xc0: void** iTable = !j9x 0x0000000000049B90
	0xc8: UDATA castClassCache = 0x000000000004B401 (308225)
	0xd0: void** jniIDs = !j9x 0x0000000000000000
	0xd8: UDATA lockOffset = 0xFFFFFFFFFFFFFFFF (-1)
	0xe0: U32 paddingForGLRCounters = 0x00000000 (0)
	0xe4: U16 reservedCounter = 0x0000 (0)
	0xe6: U16 cancelCounter = 0x0000 (0)
	0xe8: UDATA newInstanceCount = 0x00000000000003E8 (1000)
	0xf0: IDATA backfillOffset = 0x0000000000000004 (4)
	0xf8: struct J9Class* replacedClass = !j9class 0x0000000000000000
	0x100: UDATA finalizeLinkOffset = 0x0000000000000000 (0)
	0x108: struct J9Class* nextClassInSegment = !j9class 0x000000000004B600 // int
	0x110: UDATA* ramConstantPool = !j9x 0x0000000000000000
	0x118: struct J9Object** callSites = !j9x 0x0000000000000000
	0x120: struct J9Object** methodTypes = !j9x 0x0000000000000000
	0x128: struct J9Object** varHandleMethodTypes = !j9x 0x0000000000000000
	0x130: struct J9VMCustomSpinOptions* customSpinOption = !j9vmcustomspinoptions 0x0000000000000000
	0x138: void** staticSplitMethodTable = !j9x 0x0000000000000000
	0x140: void** specialSplitMethodTable = !j9x 0x0000000000000000
	0x148: struct J9JITExceptionTable* jitMetaDataList = !j9jitexceptiontable 0x0000000000000000
	0x150: struct J9Class* gcLink = !j9class 0x0000000000000000
	0x158: struct J9Class* hostClass = !j9class 0x000000000004B800 // [I
	0x160: struct J9Class* nestHost = !j9class 0x0000000000000000
	0x168: struct J9FlattenedClassCache* flattenedClassCache = !j9flattenedclasscache 0x0000000000000004
	0x170: struct J9ClassHotFieldsInfo* hotFieldsInfo = !j9classhotfieldsinfo 0x0000000000000000
}
Class name: [I
To view static fields, use !j9statics 0x000000000004B800
To view instance shape, use !j9classshape 0x000000000004B800

> !j9romclass 0x00007F5E65261260
J9ROMClass at 0x7f5e65261260 {
  Fields for J9ROMClass:
	0x0: U32 romSize = 0x000000B0 (176)
	0x4: U32 singleScalarStaticCount = 0x00000000 (0)
	0x8: J9SRP(J9UTF8) className = !j9utf8 0x00007F5E652613E4
	0xc: J9SRP(J9UTF8) superclassName = !j9utf8 0x00007F5E652613EC
	0x10: U32 modifiers = 0x00010411 (66577)
	0x14: U32 extraModifiers = 0x80400000 (2151677952)
	0x18: U32 interfaceCount = 0x00000002 (2)
	0x1c: J9SRP(J9SRP(J9UTF8)) interfaces = !j9utf8 0x00007F5E652613FE
	0x20: U32 romMethodCount = 0x00000000 (0)
	0x24: J9SRP(J9ROMMethod) romMethods = !j9rommethod 0x00007F5E65261286
	0x28: U32 romFieldCount = 0x00000000 (0)
	0x2c: J9SRP(J9ROMFieldShape) romFields = !j9romfieldshape 0x0000000000000000
	0x30: U32 objectStaticCount = 0x00000000 (0)
	0x34: U32 doubleScalarStaticCount = 0x00000000 (0)
	0x38: U32 ramConstantPoolCount = 0x00000000 (0)
	0x3c: U32 romConstantPoolCount = 0x00000000 (0)
	0x40: J9WSRP(U8) intermediateClassData = <FAULT>
	0x48: U32 intermediateClassDataLength = 0x00000000 (0)
	0x4c: U32 instanceShape = 0x00000406 (1030)
	0x50: J9SRP(U32) cpShapeDescription = <FAULT>
	0x54: J9SRP(J9UTF8) outerClassName = !j9utf8 0x0000000000000000
	0x58: U32 memberAccessFlags = 0x00000000 (0)
	0x5c: U32 innerClassCount = 0x00000000 (0)
	0x60: J9SRP(J9SRP(J9UTF8)) innerClasses = <FAULT>
	0x64: J9SRP nestHost = !j9x 0x0000000000000000
	0x68: U16 nestMemberCount = 0x0000 (0)
	0x6a: U16 unused = 0x0000 (0)
	0x6c: J9SRP nestMembers = !j9x 0x0000000000000000
	0x70: U16 majorVersion = 0x0000 (0)
	0x72: U16 minorVersion = 0x0000 (0)
	0x74: U32 optionalFlags = 0x00000000 (0)
	0x78: J9SRP(U32) optionalInfo = <FAULT>
	0x7c: U32 maxBranchCount = 0x00000000 (0)
	0x80: U32 methodTypeCount = 0x00000000 (0)
	0x84: U32 varHandleMethodTypeCount = 0x00000000 (0)
	0x88: U32 bsmCount = 0x00000000 (0)
	0x8c: U32 callSiteCount = 0x00000000 (0)
	0x90: J9SRP callSiteData = !j9x 0x0000000000000000
	0x94: U32 classFileSize = 0x00000000 (0)
	0x98: U32 classFileCPCount = 0x00000000 (0)
	0x9c: U16 staticSplitMethodRefCount = 0x0000 (0)
	0x9e: U16 specialSplitMethodRefCount = 0x0000 (0)
	0xa0: J9SRP(U16) staticSplitMethodRefIndexes = <FAULT>
	0xa4: J9SRP(U16) specialSplitMethodRefIndexes = <FAULT>
	0xa8: J9SRP varHandleMethodTypeLookupTable = !j9x 0x0000000000000000
	0xac: U32 padding = 0x00000000 (0)
}

```

Reply:
@hangshao0 we should propogate the `J9ClassHasIdentity` flag to the J9ArrayClass for consistency

Reply:
I think you just need to add it to https://github.com/eclipse-openj9/openj9/blob/master/runtime/vm/createramclass.cpp#L3127

Reply:
PR: https://github.com/eclipse-openj9/openj9/pull/13051

Reply:
> PR: #13051 

It is merged.

Reply:
The check on `isClassArray` is removed in f926136